nixos/modules/services/web-apps/firefly-iii.nix

```diff
@@ -0,0 +1,438 @@
+      inherit user;
+      inherit group;
+      phpOptions = ''
+        log_errors = on
+      '';
+      settings = {
+        "listen.mode" = "0660";
+        "listen.owner" = user;
+        "listen.group" = group;
+      } // cfg.poolConfig;
+    };
+
+    # MySQL
+    services.mysql = mkIf db.createLocally {
+      enable = true;
+      package = mkDefault pkgs.mariadb;
+      ensureDatabases = [ db.name ];
+      ensureUsers = [
+        {
+          name = db.user;
+          ensurePermissions = { "${db.name}.*" = "ALL PRIVILEGES"; };
+        }
+      ];
+    };
+
+    # Reverse proxy
+    services.nginx = {
+      enable = mkDefault true;
+      recommendedTlsSettings = true;
+      recommendedOptimisation = true;
+      recommendedGzipSettings = true;
+      virtualHosts.${cfg.hostname} = mkMerge [ cfg.nginx {
+        root = mkForce "${firefly-iii}/public";
+        locations = {
+          "/" = {
+            index = "index.php";
+            tryFiles = "$uri $uri/ /index.php?$query_string";
+          };
+          "~ \.php$".extraConfig = ''
+            fastcgi_pass unix:${config.services.phpfpm.pools."firefly-iii".socket};
+          '';
+          "~ \.(js|css|gif|png|ico|jpg|jpeg)$" = {
+            extraConfig = "expires 365d;";
+          };
+        };
+      }];
+    };
+
+    # Config
+    services.firefly-iii.config = {
+      APP_URL = cfg.appURL;
+      APP_KEY._secret = cfg.appKeyFile;
+
+      DB_CONNECTION = db.type;
+      DB_HOST = db.host;
+      DB_PORT = db.port;
+      DB_DATABASE = db.name;
+      DB_USERNAME = db.user;
+      DB_PASSWORD._secret = db.passwordFile;
+
+      MAIL_MAILER = mail.driver;
+      MAIL_HOST = mail.host;
+      MAIL_PORT = mail.port;
+      MAIL_FROM = mail.from;
+      MAIL_USERNAME = mail.user;
+      MAIL_PASSWORD._secret = mail.passwordFile;
+      MAIL_ENCRYPTION = mail.encryption;
+    };
+
+    # Set-up script
+    environment.systemPackages = [ artisan ];
+
+    systemd.services.firefly-iii-setup = {
+      description = "Preparation tasks for Firefly III";
+      before = [ "phpfpm-firefly-iii.service" ];
+      after = optional db.createLocally "mysql.service";
+      wantedBy = [ "multi-user.target" ];
+      serviceConfig = {
+        Type = "oneshot";
+        RemainAfterExit = true;
+        User = user;
+        WorkingDirectory = "${firefly-iii}";
+      };
+      path = [ pkgs.replace-secret ];
+      script =
+        let
+          isSecret  = v: isAttrs v && v ? _secret && (isString v._secret || builtins.isPath v._secret);
+          fireflyEnvVars = generators.toKeyValue {
+            mkKeyValue = flip generators.mkKeyValueDefault "=" {
+              mkValueString = v: with builtins;
+                if isInt         v then toString v
+                else if isString v then v
+                else if true  == v then "true"
+                else if false == v then "false"
+                else if isSecret v then hashString "sha256" v._secret
```

Comment:
If the `appKeyFile` is specified as path, `hashString` can not be used. I think you meant to use `hashFile`?

Reply:
@timhae , I think it's OK now, although I can't seem to remember when I got this actually fixed:

```nix
else if isSecret     v then
                  if (isString v._secret) then
                    hashString "sha256" v._secret
                  else
                    hashString "sha256" (builtins.readFile v._secret)
```